CSVPrimitiveTensor

Wendy-Ji opened the pull request:
This user tensor supports reading Scalar, Vec, and SR2 data that is sorted by component in CSV files, e.g. a displacement vector's components provided in the columns: disp_x, disp_y, disp_z. Each row of the CSV is a different instance of the tensor. The tensor can optionally be reshaped by the user to a specified batch shape.

This is a continuation of PR #287, which was split into this PR and PR #293 

Since this shares a lot of the same structure as `MultiColumnCSVScalar` (PR #293), a `CSVReader` has been created for the common functions (`parse_format`, `parse_indices`, `read_all`, `read_by_indices`). Some differences are:
- Only row-major indexing, with default batch size set to number of rows in CSV
- Error checks added for ensuring that no. of columns read from CSV = no. of elements in base size (also can help with catching an incorrectly selected tensor type)

This should close issue #154

Wendy-Ji committed:
CSVPrimitiveTensor

Wendy-Ji commented on tests/unit/user_tensors/test_CSVPrimitiveTensor.cxx:
```diff
@@ -0,0 +1,202 @@
+    REQUIRE(at::allclose(*parsed, expected));
+  }
+
+  SECTION("errors")
```
I'm not sure if I need to include this many tests, e.g. non-numeric has four different test cases for whether it's `read_all` or `read_by_indices` and whether it has a header or not. 

Also, I relabelled the test cases so they have more descriptive names instead of being called error_1, error_2, etc. Let me know if you want it changed back

Wendy-Ji commented on src/neml2/user_tensors/MultiColumnCSVScalar.cxx:
```diff
@@ -214,19 +214,27 @@ MultiColumnCSVScalar::read_all(csv::CSVReader & csv,
-      vals.push_back(field.get<double>());
-      if (nrow == 0)
-        ncol++;
+      if (no_header)
```
Realised that the `read_by_indices` had two separate error messages depending on whether there was a header or not, so that the message would output either a column name or a column index. Added the same capability to `read_all`

Wendy-Ji commented on src/neml2/user_tensors/MultiColumnCSVScalar.cxx:
```diff
@@ -254,16 +262,17 @@ MultiColumnCSVScalar::read_by_indices(csv::CSVReader & csv,
-                    csv.get_col_names()[idx],
-                    "', row ",
-                    nrow);
+                    "Non-numeric value found in CSV file at row '",
```
For consistency with the other error messages, changed the error messages to output row number first, then column

Wendy-Ji commented on src/neml2/user_tensors/CSVPrimitiveTensor.cxx:
```diff
@@ -0,0 +1,409 @@
+  auto tensor_type = utils::demangle(typeid(T).name()).substr(7);
+
+  OptionSet options = UserTensorBase::expected_options();
+  options.doc() =
```
`CSVPrimitiveTensor` shares many of the exact same functions as `MultiColumnCSVScalar`. Is this repetition alright?

hugary1995 replied:
Duplicate code will be a pain to maintain. If two functions are similar, factor out the common bits, implement it once in one translation unit, and `#include` it in the other translation unit that uses it.

For example, say `read_all` is used by both csv user tensors, there are two typical ways of avoiding duplication:
- Factor out a base class for csv readers with common functions such as `read_all` and let both csv user tensors derive from it.
- Move `read_all` outside the class to be a free function, then whenever you need to call the function, just `#include` the header that defines the function signature.

hugary1995 commented:
You can see the error message by clicking on the failed check. Most of them are build errors. You can follow what the error message suggested, i.e. insert `template` before get.
```
  135 |   const auto delim = options.get<EnumSelection>("delimiter").as<char>();
      |                              ^
      |                              template
```

You can read https://en.cppreference.com/w/cpp/language/dependent_name.html for why you need to do this. Specifically the section titled "The template disambiguator...".

Wendy-Ji committed:
CSVReader, merge changes in UserTensorBase from main

hugary1995 commented:
You can find details about the clang-tidy checks here https://applied-material-modeling.github.io/neml2/pr-preview/pr-297/clang-tidy

Wendy-Ji commented on src/neml2/user_tensors/CSVPrimitiveTensor.cxx:
```diff
@@ -0,0 +1,175 @@
+
+template <typename T>
+void
+CSVPrimitiveTensor<T>::multiply_factor(std::vector<double> & vals,
```
Don't need to modify CSV values for Scalar and Vector, but this gives unused parameter warnings if this is left empty

Wendy-Ji commented on src/neml2/user_tensors/CSVPrimitiveTensor.cxx:
```diff
@@ -0,0 +1,175 @@
+  if (indices.empty())
+  {
+    CSVTensorBase::read_all(csv, vals, nrow, ncol);
+    if (ncol != base_size_ele)
```
This error checks whether the number of columns matches the number of components needed for the tensor. `check_col` does the same thing, but `check_col` is only for if the user has specified `column_names` or `column_indices` (which gives a number of columns to compare to). For the `read_all` case, the erorr check needs `ncol`, which is obtained from reading the CSV. I'm not sure where else to put this error

Wendy-Ji commented on src/neml2/user_tensors/CSVPrimitiveTensor.cxx:
```diff
@@ -0,0 +1,177 @@
+  multiply_factor(vals, nrow, ncol);
+
+  // Convert to tensor and reshape
+  std::vector<std::vector<double>> vals_double(nrow, std::vector<double>(ncol));
```
Fixes clang-tidy, but seems like a lot of for loops

hugary1995 commented:
OMG, I didn't notice you marked this as ready for review. I swear I didn't see a notification...

Wendy-Ji commented:
> OMG, I didn't notice you marked this as ready for review. I swear I didn't see a notification...

I saw you were working on a massive PR. Congratulations on finishing that.

Does this PR need any changes to be merged?

hugary1995 committed:
minor cleanup; remove batch_shape

hugary1995 commented:
I just rebased your branch on main, and did some minor cleanup. Take a look and let me know if the changes look reasonable to you.

One notable change is that I removed the `batch_shape` option. As now all user tensors support general shape manipulations, so there's no need to have a specialized implementation for the csv tensors. The relevant options are `shape_manipulations` and `shape_manipulation_args` which support a wide variety of reshaping operations.

Wendy-Ji commented:
> I just rebased your branch on main, and did some minor cleanup. Take a look and let me know if the changes look reasonable to you.
> 
> One notable change is that I removed the `batch_shape` option. As now all user tensors support general shape manipulations, so there's no need to have a specialized implementation for the csv tensors. The relevant options are `shape_manipulations` and `shape_manipulation_args` which support a wide variety of reshaping operations.

Looks good to me. I just wanted to check, would it be useful to still keep the error that tells the user the shape doesn't match the number of values read from the CSV?

Or does `shape_manipulations` already throw a similar error.

hugary1995 commented:
Yeah, there will be a torch error thrown, and from the backtrace you can see it's coming from manipulating shapes. If there's need, we could later improve the error message to be more specific.

hugary1995 commented:
Thanks for your contribution, as always :)
